google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,202 @@
+	logger.Infof("running guest consistent snapshot script at: %s", path)
+
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return -1, sspb.AgentErrorCode_SCRIPT_NOT_FOUND
+	}
+
+	execResult := runCmdOutputWithTimeout(config.timeout, path, disks)
+
+	if execResult.code == 124 {
+		return execResult.code, sspb.AgentErrorCode_SCRIPT_TIMED_OUT
+	}
+
+	if execResult.code != 0 {
+		return execResult.code, sspb.AgentErrorCode_UNHANDLED_SCRIPT_ERROR
+	}
+
+	return execResult.code, sspb.AgentErrorCode_NO_ERROR
+}
+
+func listenForSnapshotRequests(address string, requestChan chan<- *sspb.GuestMessage) {
+	for {
+		// Start hanging connection on server that feeds to channel
+		logger.Infof("attempting to connect to snapshot service at %s", address)
+		conn, err := grpc.Dial(address, grpc.WithInsecure())
+		if err != nil {
+			logger.Errorf("failed to connect: %v", err)
+			continue
+		}
+
+		c := sspb.NewSnapshotServiceClient(conn)
+		ctx, cancel := context.WithCancel(context.Background())
+		guestReady := sspb.GuestReady{
+			RequestServerInfo: false,
+		}
+		r, err := c.CreateConnection(ctx, &guestReady)
+		if err != nil {
+			logger.Errorf("error creating connection: %v", err)
+			continue
+		}
+		for {
+			request, err := r.Recv()
+			if err != nil {
+				logger.Errorf("error reading snapshot request: %v", err)
+				break
+			}
+			logger.Infof("received snapshot request")
+			requestChan <- request
+		}
+		cancel()
+	}
+}
+
+func getSnapshotResponse(guestMessage *sspb.GuestMessage) *sspb.SnapshotResponse {
+	switch {
+	case guestMessage.GetSnapshotRequest() != nil:
+		request := guestMessage.GetSnapshotRequest()
+		response := &sspb.SnapshotResponse{
+			OperationId: request.GetOperationId(),
+			Type:        request.GetType(),
+		}
+
+		config, err := getSnapshotConfig()
+		if err != nil {
+			response.AgentReturnCode = sspb.AgentErrorCode_INVALID_CONFIG
+			return response
+
+		}
+
+		var url string
+		switch request.GetType() {
+		case sspb.OperationType_PRE_SNAPSHOT:
+			logger.Infof("handling pre snapshot request for operation id %d", request.GetOperationId())
+			_, found := seenPreSnapshotOperationIds.Get(request.GetOperationId())
+			if found {
+				logger.Infof("duplicate pre snapshot request operation id %d", request.GetOperationId())
+				return nil
+			}
+			seenPreSnapshotOperationIds.Add(request.GetOperationId(), request.GetOperationId())
+			url = config.preSnapshotScriptURL
+		case sspb.OperationType_POST_SNAPSHOT:
+			logger.Infof("handling post snapshot request for operation id %d", request.GetOperationId())
+			_, found := seenPostSnapshotOperationIds.Get(request.GetOperationId())
+			if found {
+				logger.Infof("duplicate post snapshot request operation id %d", request.GetOperationId())
+				return nil
+			}
+			seenPostSnapshotOperationIds.Add(request.GetOperationId(), request.GetOperationId())
+			url = config.postSnapshotScriptURL
+		default:
+			logger.Errorf("unhandled operation type %d", request.GetType())
```

Comment:
Log lines should be capitalized and end with punctuation (here and elsewhere)

Reply:
Fixed!